syntax = "proto3";

import public "attach_network_interface_request.proto";
import public "attach_volume_request.proto";
import public "create_ipv4_subnet_request.proto";
import public "create_network_request.proto";
import public "create_vm_request.proto";
import public "create_vm_snapshot_request.proto";
import public "create_volume_request.proto";
import public "delete_ipv4_subnet_request.proto";
import public "delete_network_request.proto";
import public "delete_vm_request.proto";
import public "delete_vm_snapshot_request.proto";
import public "delete_volume_request.proto";
import public "detach_network_interface_request.proto";
import public "detach_volume_request.proto";
import public "update_ipv4_subnet_request.proto";
import public "update_network_interface_request.proto";
import public "update_network_request.proto";
import public "update_vm_power_state_request.proto";
import public "update_vm_request.proto";

import public "enums/network_type.proto";
import public "enums/vm_power_state.proto";
import "vlan_network_setting.proto";

message Request {
  // request specific fields into here

  oneof message {
    CreateVMRequest CreateVMRequest = 11;
    UpdateVMRequest UpdateVMRequest = 12;
    DeleteVMRequest DeleteVMRequest = 13;

    UpdateVMPowerStateRequest UpdateVMPowerStateRequest = 14;

    CreateVolumeRequest CreateVolumeRequest = 15;
    DeleteVolumeRequest DeleteVolumeRequest = 16;
    AttachVolumeRequest AttachVolumeRequest = 17;
    DetachVolumeRequest DetachVolumeRequest = 18;

    CreateNetworkRequest CreateNetworkRequest = 19;
    UpdateNetworkRequest UpdateNetworkRequest = 20;
    DeleteNetworkRequest DeleteNetworkRequest = 21;

    AttachNetworkInterfaceRequest AttachNetworkInterfaceRequest = 22;
    UpdateNetworkInterfaceRequest UpdateNetworkInterfaceRequest = 23;
    DetachNetworkInterfaceRequest DetachNetworkInterfaceRequest = 24;

    CreateIPv4SubnetRequest CreateIPv4SubnetRequest = 25;
    UpdateIPv4SubnetRequest UpdateIPv4SubnetRequest = 26;
    DeleteIPv4SubnetRequest DeleteIPv4SubnetRequest = 27;
  }
}
